base/shell/explorer/CTrayShowDesktopButton.cpp

```diff
@@ -0,0 +1,405 @@
+
+#define SHOW_DESKTOP_TIMER_ID 999
+#define SHOW_DESKTOP_TIMER_INTERVAL 200
+
+VOID CTrayShowDesktopButton::StartHovering()
+{
+    if (m_bHovering)
+        return;
+
+    m_bHovering = TRUE;
+    Invalidate(TRUE);
+
+    SetTimer(SHOW_DESKTOP_TIMER_ID, SHOW_DESKTOP_TIMER_INTERVAL, NULL);
+
+    HWND taskbarWnd;
+    if (GetTaskbar(&taskbarWnd))
+        ::PostMessage(taskbarWnd, WM_NCPAINT, 0, 0);
+}
+
+LRESULT CTrayShowDesktopButton::OnMouseMove(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    StartHovering();
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnTimer(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    if (wParam != SHOW_DESKTOP_TIMER_ID || !m_bHovering)
+        return 0;
+
+    POINT pt;
+    ::GetCursorPos(&pt);
+    if (!PtInButton(&pt)) // The end of hovering?
+    {
+        m_bHovering = FALSE;
+        KillTimer(SHOW_DESKTOP_TIMER_ID);
+        Invalidate(TRUE);
+        
+        HWND taskbarWnd;
+        if (GetTaskbar(&taskbarWnd))
+            ::PostMessage(taskbarWnd, WM_NCPAINT, 0, 0);
+    }
+
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnDestroy(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    if (m_hTheme)
+    {
+        CloseThemeData(m_hTheme);
+        m_hTheme = NULL;
+    }
+    return 0;
+}
+
+VOID CTrayShowDesktopButton::OnDraw(HDC hdc, LPRECT prc)
+{
+    RECT rc = { prc->left, prc->top, prc->right, prc->bottom };
+    LPRECT lpRc = &rc;
+    HBRUSH hbrBackground = NULL;
+    if (m_hTheme)
+    {
+        HTHEME theme;
+        int part = 0;
+        int state = 0;
+        if (m_drawWithDedicatedBackground)
+        {
+            theme = m_hTheme;
+            if (m_bPressed)
+                state = PBS_PRESSED;
+            else if (m_bHovering)
+                state = PBS_HOT;
+            else
+                state = PBS_NORMAL;
+        }
+        else
+        {
+            part = TP_BUTTON;
+            theme = m_hFallbackTheme;
+
+            if (m_bPressed)
+                state = TS_PRESSED;
+            else if (m_bHovering)
+                state = TS_HOT;
+            else
+                state = TS_NORMAL;
+            
+            if (IsHorizontal)
+                rc.right -= m_inset.cx;
+            else
+                rc.bottom -= m_inset.cy;
+        }
+        
+        if (::IsThemeBackgroundPartiallyTransparent(theme, part, state))
+            ::DrawThemeParentBackground(m_hWnd, hdc, NULL);
+
+        ::DrawThemeBackground(theme, hdc, part, state, lpRc, lpRc);
+    }
+    else
+    {
+        hbrBackground = ::GetSysColorBrush(COLOR_3DFACE);
+        ::FillRect(hdc, lpRc, hbrBackground);
+
+        if (m_bPressed || m_bHovering)
+        {
+            UINT edge = m_bPressed ? BDR_SUNKENOUTER : BDR_RAISEDINNER;
+            DrawEdge(hdc, lpRc, edge, BF_RECT);
+        }
+    }
+
+    if (m_highContrastMode || !m_drawWithDedicatedBackground)
+    {
+        /* Prepare to draw icon */
+
+        int iconSize = 16;
```

Comment:
Should this be derived from a setting?

Reply:
...good question. The question hadn't even occurred to me...a frightening oversight.

Seems I'll have to investigate.

Reply:
@learn-more Apparently it is...but there's DPI scaling stuff involved or something, so that's...not really within the realm of possibility at the moment...right?

Reply:
I was more thinking along the lines of `GetSystemMetrics(SM_CXSMICON)` or the likes,
or some taskbar specific setting.

Reply:
@learn-more Well apparently that's not it - according to Hermes, Windows draws the icon via `DrawIconEx`, passing 0 for the width and height and `DI_NORMAL` for `diFlags`.

If I've understood the docs for `DrawIconEx` correctly, that means...the icon resource itself should govern the size it is drawn at? if so, then...I guess I'd have to do the same to determine how to properly center the icon as well, right?

Also, since I'm not 100% sure I even understood the docs correctly, here's the page in question: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-drawiconex

Reply:
> If I've understood the docs for DrawIconEx correctly, that means...the icon resource itself should govern the size it is drawn at? if so, then...I guess I'd have to do the same to determine how to properly center the icon as well, right?

Reading the page you linked, indeed.

We have 2 implementations of that:
https://github.com/reactos/reactos/blob/7dd33088282e7029f3a81f939dbd24835954f953/dll/win32/comctl32/static.c#L88
and:
https://github.com/reactos/reactos/blob/7dd33088282e7029f3a81f939dbd24835954f953/win32ss/user/user32/windows/cursoricon.c#L74